osgp/shared/shared/src/main/java/org/opensmartgridplatform/shared/security/providers/HsmEncryptionProvider.java

```diff
@@ -0,0 +1,108 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.opensmartgridplatform.shared.security.providers;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.NoSuchProviderException;
+import java.security.UnrecoverableKeyException;
+import java.security.cert.CertificateException;
+import java.security.spec.AlgorithmParameterSpec;
+import java.util.Arrays;
+
+import javax.crypto.Cipher;
+import javax.crypto.NoSuchPaddingException;
+import javax.crypto.spec.IvParameterSpec;
+
+import org.apache.commons.codec.binary.Hex;
+import org.opensmartgridplatform.shared.exceptionhandling.EncrypterException;
+import org.opensmartgridplatform.shared.security.EncryptedSecret;
+import org.opensmartgridplatform.shared.security.EncryptionProviderType;
+import org.opensmartgridplatform.shared.security.Secret;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class HsmEncryptionProvider extends AbstractEncryptionProvider implements EncryptionProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(HsmEncryptionProvider.class);
+
+    private static final int KEY_LENGTH = 16;
+    private static final String ALGORITHM = "AES/CBC/NoPadding";
+    private static final String PROVIDER = "nCipherKM";
+    private static final String TYPE = "ncipher.sworld";
+    private static final byte[] IV = new byte[KEY_LENGTH];
+
+    private final KeyStore keyStore;
+
+    public HsmEncryptionProvider(File keyStoreFile) {
+        try {
+            super.setKeyFile(keyStoreFile);
+            this.keyStore = KeyStore.getInstance(TYPE, PROVIDER);
+            FileInputStream fIn = new FileInputStream(keyStoreFile);
+            this.keyStore.load(fIn, null);
+        } catch (CertificateException | NoSuchAlgorithmException | NoSuchProviderException | IOException | KeyStoreException e) {
+            throw new EncrypterException("Could not read keystore", e);
+        }
+    }
+
+    public Secret decrypt(EncryptedSecret secret, String keyReference) {
+
+        Secret decryptedSecret = super.decrypt(secret, keyReference);
+
+        byte[] decryptedSecretBytes = decryptedSecret.getSecret();
+
+        if (decryptedSecretBytes.length > KEY_LENGTH) {
+
+            byte[] truncatedDecryptedSecretBytes = Arrays.copyOfRange(decryptedSecretBytes, 0,
+                    decryptedSecretBytes.length-16);
+
+            LOGGER.trace("Truncating decrypted key from " + Hex.encodeHexString(decryptedSecretBytes) + " to " +
+                            Hex.encodeHexString(truncatedDecryptedSecretBytes));
+
+            decryptedSecret = new Secret(truncatedDecryptedSecretBytes);
+        }
+
+        return decryptedSecret;
+    }
+
+    protected Cipher getCipher() {
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove the declaration of thrown exception 'org.opensmartgridplatform.shared.exceptionhandling.EncrypterException' which is a runtime exception. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://ci.opensmartgridplatform.org/sonarqube/coding_rules#rule_key=squid%3ARedundantThrowsDeclarationCheck)